Cloud-native object storage module

aluminiumgeek opened the pull request:
Related issue #70

aluminiumgeek committed:
Add S3 storage module

aluminiumgeek committed:
Provide ability to switch to cloud native storage instead of minio on cluster/storage provider initialization

aluminiumgeek committed:
Add gcloud storage module, +tests

akariv commented on ckan_cloud_operator/providers/storage/manager.py:
```diff
@@ -1,9 +1,46 @@
+            secret_name=CONFIG_NAME,
+            interactive=interactive
+        )
+        if config_manager.get('use-cloud-native-storage', secret_name=CONFIG_NAME):
```
Why not move these following initializations to the matching {s3/gcp}/manager.py initialize functions?

akariv commented on ckan_cloud_operator/providers/storage/s3/manager.py:
```diff
@@ -0,0 +1,92 @@
+        logs.warning(f'No bucket found for the instance "{instance_id}" on S3. Skipping.')
+        return
+
+    instance = kubectl.get(f'ckancloudckaninstance {instance_id}')
```
use the proper parameters please...

aluminiumgeek replied:
@akariv what do you mean?

akariv commented on ckan_cloud_operator/providers/storage/s3/manager.py:
```diff
@@ -0,0 +1,92 @@
+    }
+
+
+def list_buckets():
```
seems that this shouldn't be part of s3 as it's provider agnostic, no?

akariv replied:
(same for get_bucket above)

aluminiumgeek replied:
It's not provider agnostic. `ckancloudckaninstance` entry contains slightly different structure for providers. For example, s3 will have
```
spec:
  bucket:
    s3: s3://example
```
Gcloud will have
```
spec:
  bucket:
    gcloud: gs://example2
```

akariv committed:
Merge branch 'master' into object-storage

akariv commented:
Added a few comments @aluminiumgeek .

Please proceed with integrating with the CKAN flow.

aluminiumgeek committed:
Move storage region config to module initialize function

aluminiumgeek committed:
Merge branch 'object-storage' of github.com:datopian/ckan-cloud-operator into object-storage

aluminiumgeek committed:
Use explicit parameters for initialize()

aluminiumgeek committed:
- Wrap storage module with app instance creation flow
- Ask for AWS user access key/secret for S3 storage module
- Change return value of create_bucket()

aluminiumgeek committed:
Set useCloudStorage in instance spec

aluminiumgeek commented:
Latest update relates to https://github.com/ViderumGlobal/ckan-cloud-helm/pull/14

akariv committed:
Merge branch 'master' into object-storage

akariv commented on ckan_cloud_operator/providers/ckan/instance/manager.py:
```diff
@@ -39,14 +41,23 @@ def create(instance_type, instance_id=None, instance_name=None, values=None, val
     values['id'] = instance_id
+
+    use_cloud_storage = config_manager.get('use-cloud-native-storage', secret_name=CONFIG_NAME)
+    values['useCloudStorage'] = use_cloud_storage
```
why override this value and not use the value in the spec?

akariv replied:
If this value is true, we can verify`use-cloud-native-storage` is indeed true.

aluminiumgeek replied:
done

akariv commented on ckan_cloud_operator/providers/ckan/instance/manager.py:
```diff
@@ -39,14 +41,23 @@ def create(instance_type, instance_id=None, instance_name=None, values=None, val
     if instance_name:
         set_name(instance_id, instance_name, dry_run=dry_run)
+
+    if use_native_storage:
```
we need to create a bucket only if values['useCloudStorage'] is set

aluminiumgeek replied:
done

akariv commented on ckan_cloud_operator/providers/ckan/instance/manager.py:
```diff
@@ -64,6 +75,23 @@ def update(instance_id_or_name, override_spec=None, persist_overrides=False, wai
+                    literal.append(f'--from-literal={key}={value}')
+                literal = ' '.join(literal)
+
+                kubectl.call(f'create secret generic bucket-credentials {literal}', namespace=instance_id)
```
please use config manager instead of calling kubectl create secret directly.

aluminiumgeek replied:
done

akariv commented on ckan_cloud_operator/providers/ckan/instance/manager.py:
```diff
@@ -284,6 +312,32 @@ def set_name(instance_id, instance_name, dry_run=False):
+    resource = crds_manager.get(INSTANCE_NAME_CRD_SINGULAR, name=instance_name, required=False)
+
+    cluster_provider_id = cluster_manager.get_provider_id()
+    if cluster_provider_id == 'gcloud':
```
code duplication...

aluminiumgeek replied:
done

akariv commented on ckan_cloud_operator/providers/ckan/instance/manager.py:
```diff
@@ -284,6 +312,32 @@ def set_name(instance_id, instance_name, dry_run=False):
+
+    storage_provider = get_provider(default=None, provider_id=provider_id)
+    bucket = storage_provider.create_bucket(instance_id, exists_ok=True, dry_run=dry_run)
+    resource['spec']['bucket'] = {
```
let's be more verbose here - ckanStorageBucket instead of just 'bucket'

aluminiumgeek replied:
done

aluminiumgeek committed:
- Do not override useCloudStorage value
- create bucket-credentials secret with config manager
- refactor getting of storage provider id
- bucket -> ckanStorageBucket in spec

akariv commented:
@aluminiumgeek 
Looks good but please make sure the tests pass.

aluminiumgeek committed:
Default to minio

aluminiumgeek committed:
Set default.config.ckan-cloud-provider-storage-minio.disk-size-gb in interactive file

aluminiumgeek committed:
Set default.config.ckan-cloud-provider-storage-minio.router-name in interactive file

aluminiumgeek committed:
Fix for AssertionError

aluminiumgeek committed:
Merge branch 'master' into object-storage

aluminiumgeek committed:
Add storage-config to interactive.yaml

aluminiumgeek committed:
Add `selector` to spec

aluminiumgeek committed:
change router name

aluminiumgeek committed:
Change variable name

aluminiumgeek committed:
Default get_provider_id to minio

aluminiumgeek committed:
Add storage-config secret to main.tf

aluminiumgeek committed:
Add log statemtnt about chosen provider id

aluminiumgeek committed:
Add s3 and gcloud storage managers to submodule ids

aluminiumgeek committed:
Add is_secret parameter

aluminiumgeek committed:
Add storage-config vars to main.tf

aluminiumgeek committed:
Freeze python-dateutil version

aluminiumgeek committed:
Merge branch 'master' into object-storage

akariv commented:
@aluminiumgeek 

Please enable  cloud storage on the cluster and instance levels when testing on CI.

aluminiumgeek committed:
Enable cloud native storage in main.tf

akariv commented:
Hey @aluminiumgeek , don't forget you also need to enable cloud storage in the instance values too...

aluminiumgeek committed:
Generate and store unique bucket name; add useCloudStorage to values.aws.yaml

aluminiumgeek committed:
Clean up empty buckets

aluminiumgeek committed:
Fix NameError

akariv committed:
Merge branch 'master' into object-storage

akariv commented on ckan_cloud_operator/providers/storage/s3/manager.py:
```diff
@@ -0,0 +1,128 @@
+    default_zone = aws_config_get('aws-default-region', is_secret=True)
+    assert default_zone, 'No cluster region specified.'
+
+    if interactive and not dry_run:
```
Why do we need separate credentials and region?

akariv replied:
and `storage-region` is not even used anywhere...

aluminiumgeek replied:
@akariv separate credentials is used if you want to use separate AWS account with permissions limited only to S3.
`storage-region` is used to override default storage region (I updated the `create_bucket()` function)

akariv commented on ckan_cloud_operator/providers/ckan/instance/manager.py:
```diff
@@ -70,6 +82,19 @@ def update(instance_id_or_name, override_spec=None, persist_overrides=False, wai
         pre_update_hook_data = deployment_manager.pre_update_hook(instance_id, instance_type, instance, override_spec,
                                                                   skip_route)
+
+        if config_manager.get('use-cloud-native-storage', secret_name=CONFIG_NAME):
```
why not test for `use_cloud_storage here`?

aluminiumgeek replied:
Done

aluminiumgeek committed:
Use `storage-region`; check if instance has storage

aluminiumgeek committed:
Merge branch 'object-storage' of github.com:datopian/ckan-cloud-operator into object-storage

akariv commented on ckan_cloud_operator/providers/ckan/instance/manager.py:
```diff
@@ -83,7 +83,9 @@ def update(instance_id_or_name, override_spec=None, persist_overrides=False, wai
                                                                   skip_route)
 
-        if config_manager.get('use-cloud-native-storage', secret_name=CONFIG_NAME):
+        use_cloud_storage = instance['spec'].get('ckanStorageBucket', {}).get(get_storage_provider_id()) and config_manager.get('use-cloud-native-storage', secret_name=CONFIG_NAME)
```
looks like we need to make a function out of it.

aluminiumgeek replied:
Better reuse `bucket_credentials` var. Done

akariv commented on sample-values/values.aws.yaml:
```diff
@@ -55,3 +55,4 @@ terminationGracePeriodSeconds: 30
 useCentralizedInfra: true
 usePersistentVolumes: true
 withSansSSL: true
+useCloudStorage: true
```
which ckan-cloud-helm release introduced the use of this parameter?

aluminiumgeek replied:
helm ckan v0.0.15

akariv replied:
Then I'm guessing you should use that version here, no? Otherwise this line doesn't have any effect.

aluminiumgeek replied:
Ah, sure :)
Done.

aluminiumgeek committed:
Reuse bucket_credentials var

aluminiumgeek committed:
Use v0.0.15 ckan helm chart version

akariv commented:
@aluminiumgeek can you please rebase from master and push an '#aws' commit so we can see it passes on AWS as well?

aluminiumgeek committed:
Empty commit to build #aws

aluminiumgeek committed:
Bump ckanHelmChartVersion #aws

aluminiumgeek committed:
Bump ckanHelmChartVersion to v0.0.17 #aws

akariv commented on terraform/aws/destroy_aws.sh:
```diff
@@ -1,4 +1,5 @@
 #!/bin/sh
 for vid in `aws ec2 describe-volumes | jq '.Volumes[] | .VolumeId' -r` ; do echo $vid ; aws ec2 delete-volume --volume-id "$vid"; done
 for lb in `aws elb describe-load-balancers | jq '.LoadBalancerDescriptions[] | .LoadBalancerName' -r` ; do echo $lb ; aws elb delete-load-balancer --load-balancer-name "$lb"; done
+aws s3 ls | cut -d" " -f 3 | xargs -I{} aws s3 rb s3://{}
```
@aluminiumgeek please filter out buckets that their name contains the string `persistent`

aluminiumgeek replied:
Done

aluminiumgeek committed:
Filter out persistent buckets

akariv committed:
Merge branch 'master' into object-storage

akariv committed:
Merge branch 'master' into object-storage
